homeassistant/components/mysensors.py

```diff
@@ -252,115 +367,181 @@ def gw_start(event):
+        validated = validate_child(msg.gateway, msg.node_id, child)
+        for platform, dev_ids in validated.items():
+            devices = get_mysensors_devices(hass, platform)
+            for idx, dev_id in enumerate(list(dev_ids)):
+                if dev_id in devices:
+                    dev_ids.pop(idx)
+                    signals.append(SIGNAL_CALLBACK.format(*dev_id))
+            if dev_ids:
+                discover_mysensors_platform(hass, platform, dev_ids)
+        for signal in set(signals):
+            # Only one signal per device is needed.
+            # A device can have multiple platforms, ie multiple schemas.
+            # FOR LATER: Add timer to not signal if another update comes in.
+            dispatcher_send(hass, signal)
+        end = timer()
+        if end - start > 0.1:
             _LOGGER.debug(
-                "Update: node %s, child %s sub_type %s",
-                msg.node_id, msg.child_id, msg.sub_type)
-            for callback in self.platform_callbacks:
-                callback(self, msg)
+                "Callback for node %s child %s took %.3f seconds",
+                msg.node_id, msg.child_id, end - start)
+    return mysensors_callback
 
-        return node_update
 
+def get_mysensors_name(gateway, node_id, child_id):
+    """Return a name for a node child."""
+    return '{} {} {}'.format(
+        gateway.sensors[node_id].sketch_name, node_id, child_id)
 
-class MySensorsDeviceEntity(object):
-    """Representation of a MySensors entity."""
+
+def get_mysensors_gateway(hass, gateway_id):
+    """Return gateway."""
+    if MYSENSORS_GATEWAYS not in hass.data:
+        hass.data[MYSENSORS_GATEWAYS] = {}
+    gateways = hass.data.get(MYSENSORS_GATEWAYS)
+    return gateways.get(gateway_id)
+
+
+def setup_mysensors_platform(
+        hass, domain, discovery_info, device_class, *args, **kwargs):
+    """Set up a mysensors platform."""
+    # Only act if called via mysensors by discovery event.
+    # Otherwise gateway is not setup.
+    if not discovery_info:
+        return
+    new_devices = {}
+    new_dev_ids = discovery_info[ATTR_DEVICES]
+    for dev_id in new_dev_ids:
+        devices = get_mysensors_devices(hass, domain)
+        if dev_id in devices:
+            continue
+        gateway_id, node_id, child_id, value_type = dev_id
+        gateway = get_mysensors_gateway(hass, gateway_id)
+        if not gateway:
+            continue
+        device_class_copy = device_class
+        if isinstance(device_class, dict):
+            child = gateway.sensors[node_id].children[child_id]
+            s_type = gateway.const.Presentation(child.type).name
+            device_class_copy = device_class[s_type]
+        name = get_mysensors_name(gateway, node_id, child_id)
+
+        # python 3.4 cannot unpack inside tuple, but combining tuples works
+        args_copy = args + (gateway, node_id, child_id, name, value_type)
+        devices[dev_id] = device_class_copy(*args_copy, **kwargs)
```

Comment:
Oh, see that you use it here. I rather have that you pass a specific list/dictionary as arguments if this is necessary, like `device_args=[], device_kwargs={}`, otherwise you can't ever pass keyword args to the device that exist as args in `setup_mysensors_platform`.